test/devicecontroller/DefaultDeviceController.test.ts

```diff
@@ -1037,25 +1035,300 @@ describe('DefaultDeviceController', () => {
       expect(handleEventSpy.called).to.be.true;
     });
 
-    it('throws error when choosing VideoTransformDevice', async () => {
-      class NoOpVideoTransformDevice implements VideoTransformDevice {
-        stop(): Promise<void> {
-          throw new Error('Method not implemented.');
+    it('can fail to choose VideoTransformDevice when permission is denied', done => {
+      domMockBehavior.getUserMediaSucceeds = false;
+      domMockBehavior.getUserMediaResult = UserMediaState.PermissionDeniedError;
+      domMockBehavior.asyncWaitMs = 1500;
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, stringDeviceId, [processor]);
+      deviceController
+        .chooseVideoInputDevice(device)
+        .then(() => {})
+        .catch(e => {
+          expect(e.message).to.include('Permission denied by user');
+          done();
+        });
+    });
+
+    it('can choose same VideoTransformDevice twice', done => {
+      domMockBehavior.getUserMediaSucceeds = true;
+      const mockVideoStream = new MediaStream();
+      // @ts-ignore
+      mockVideoStream.id = 'sample';
+      // @ts-ignore
+      const mockVideoTrack = new MediaStreamTrack('test', 'video');
+      mockVideoStream.addTrack(mockVideoTrack);
+      domMockBehavior.createElementCaptureStream = mockVideoStream;
+      domMockBehavior.mediaStreamTrackSettings = {
+        width: 0,
+        height: 0,
+        deviceId: 'test-same-device',
+        facingMode: 'user',
+      };
+      const processor = new NoOpVideoFrameProcessor();
+
+      deviceController.chooseVideoInputDevice('test-same-device').then(() => {});
+      const device = new DefaultVideoTransformDevice(logger, 'test-same-device', [processor]);
+
+      deviceController.chooseVideoInputDevice(device).then(() => {});
+
+      new TimeoutScheduler(200).start(() => {
+        deviceController.chooseVideoInputDevice(device);
+      });
+
+      new TimeoutScheduler(250).start(() => {
+        deviceController.chooseVideoInputDevice(null);
+      });
+
+      new TimeoutScheduler(300).start(() => {
+        device.stop();
+        done();
+      });
```

Comment:
I don't really know what you're trying to exercise with this test. You're setting up two calls to race, but reducing the likelihood of them racing by putting a constructor call in the middle. You're then triggering two delayed operations 50ms apart, and stopping the device after, but there's no guarantee there of ordering at all — the first once might take 500ms to run, and will race with the second, and the stop might complete before the previous two calls. What are you testing here?

Reply:
did a pass to remove all the TimeoutScheduler, this test is just try to choose same device twice.